File: src/model/gpu_model.cc
#include <bitset>
#include <stdexcept>
#include <assert.h>

#include "common/include/defines.h"
#include "gpu_model.h"
#include "support.h"


GPU::GPU(unsigned int _gpu_id, std::string _name) {
	this->gpu_id = _gpu_id;
	this->name = _name;
	this->util = 0.0f;
	if (name == "t400") {
		this->max_sm = T400::max_sm;
		this->total_sm = T400::total_sm;
		this->power_static = T400::power_static;
		this->power_idle = T400::power_idle;
		this->power_dynamic = T400::power_dynamic;
		// this->ghd = T400::ghd;
		// this->gdh = T400::gdh;
	}else if (name == "rtx3070") {
		this->max_sm = RTX3070::max_sm;
		this->total_sm = RTX3070::total_sm;
		this->power_static = RTX3070::power_static;
		this->power_idle = RTX3070::power_idle;
		this->power_dynamic = RTX3070::power_dynamic;
		// this->ghd = RTX3070::ghd;
		// this->gdh = RTX3070::gdh;
	}

	this->total_sm_map = (1 << this->total_sm) - 1;
	this->act_sm = 0;
	this->act_sm_map = 0x0;
	this->slots.reserve(2);
	for (int i = 0; i < partition_num; i++) {
		this->slots.push_back(GPU_JobSlot());
	}
}

GPU::~GPU() {}

int GPU::assign_slot(Job *__job, int sid) {
	if (sid != -1) { // assigned to a specific slot
		if (this->slots[sid].empty_()) {
			this->slots[sid].assign_job_slot(__job);
			add_sm_into_use(__job->get_sm_map());
			return sid;
		}else {
			std::cout << "Assigning an anavailable GPU slot! Automatically select one instead." << std::endl;
		}
	}
	for (int i = 0; i < partition_num; i++) {
		if (this->slots[i].empty_()) {
			this->slots[i].assign_job_slot(__job);
			add_sm_into_use(__job->get_sm_map());
			return i;
		}
	}
}

int GPU::assign_slot_tp(Job *__job, float tp) {
	bool failed = true;
	for (int i = 0; i < partition_num; i++) {
		if (this->slots[i].empty_tp(tp)) {
			this->slots[i].push_to_waitlist(*__job);
			return i;
		}
	}
	if (failed) {
		throw std::runtime_error("int GPU::assign_slot_tp(Job *__job, float tp) failed.");
	}
}

void GPU::reset_slot(int slot_id) {
	rev_sm_from_use(this->slots[slot_id].sm_map);
	this->slots[slot_id].delete_job_slot();
}

void GPU::reset_slot_waitlist(int slot_id) {
	this->slots[slot_id].clear_waitlist();
}

/* when the GPU is partially occupied, get the slot that is active */
unsigned int GPU::get_active_slot() {
	assert(get_status() == ACTIVE);
	if (!slots[0].empty_()) 
		return 0;
	else if (!slots[1].empty_()) 
		return 1;
}

/* when the gpu is full */
unsigned int GPU::get_faster_slot() {
	assert(get_status() == FULL);
	unsigned int slot_id = -1;
	if (!this->slots[0].empty_() && !this->slots[1].empty_()) {	// (1) two running jobs
		if (this->slots[0].job->gpu_end_time < this->slots[1].job->gpu_end_time)
			slot_id = 0;
		else
			slot_id = 1;
	} else {  // (2) one job is using all SMs, return it
		if (!this->slots[0].empty_()) slot_id = 0;
		else slot_id = 1;
	}
	return slot_id;
}

float GPU::calc_power_static() {
	return this->power_static;
}

float GPU::calc_power_idle() {
	if (get_status() == IDLE) return 0;
	else return (this->total_sm - this->act_sm) * this->power_idle;
}

float GPU::calc_power_total() {
	if (get_status() == IDLE)
		return calc_power_static();
	else {
		float power = 0;
		power += slots[0].calc_power_d(power_dynamic, slots[0].job) + slots[1].calc_power_d(power_dynamic, slots[1].job);
		power += calc_power_idle() + calc_power_static();
		return power;
	}
}

unsigned int GPU::get_sm_slot_tp(float tp, unsigned int slot_id) {
	return this->slots[slot_id].get_sm_tp(tp);	
}

unsigned int GPU::get_act_sm_tp(float tp) {
	unsigned int sm0, sm1;
	unsigned int ret = 0;
	for (int i = 0; i < this->slots.size(); i++) {
		ret += this->slots[i].get_sm_tp(tp);
	}
	return ret;
}

uint64_t GPU::gen_sm_map(unsigned int sm) {
	uint64_t ret = 0;
	uint64_t usable_sm_map = total_sm_map & (~this->act_sm_map);
	int ex = 0;

	int cnt = 0;
	while (sm > 0) {
		if (usable_sm_map & 1) {
			sm--;
			ret |= 1 << cnt;
		}
		usable_sm_map >>= 1;
		cnt++;
	}
	return ret;
}

void GPU::add_sm_into_use(uint64_t _sm_map) {
	this->act_sm_map |= _sm_map;
	this->act_sm = get_sm_by_map(this->act_sm_map);
}

void GPU::rev_sm_from_use(uint64_t _sm_map) {
	this->act_sm_map &= ~_sm_map;
	this->act_sm = get_sm_by_map(this->act_sm_map);
}

File: src/model/gpu_model.py
from typing import List

from src.model.gpu_slot_model import GPU_JobSlot
from common.define import *
from src.model.support import dict_rtx3070, dict_t400
from src.model.job_model import Job

class GPU:
	def __init__(self, _gpu_id: int, name: str):
		self.name = name
		self.gpu_id = _gpu_id
		self.util: float = 0.0
		self.total_sm: int = sm_limit[_gpu_id]
		if self.name == "t400" or self.name == "T400":
			self.max_sm = dict_t400["max_sm"]
			self.power_static = dict_t400["power_static"]
			self.power_idle = dict_t400["power_idle"]
			self.power_dynamic = dict_t400["power_dynamic"]
		elif self.name == "rtx3070" or self.name == "RTX3070":
			self.max_sm = dict_rtx3070["max_sm"]
			self.power_static = dict_rtx3070["power_static"]
			self.power_idle = dict_rtx3070["power_idle"]
			self.power_dynamic = dict_rtx3070["power_dynamic"]

		self.act_sm: int = 0
		self.slots = [GPU_JobSlot(0), GPU_JobSlot(1)]

	def get_status(self) -> gpu_status_t:
		if self.slots[0].empty_() and self.slots[1].empty_():
			return gpu_status_t.IDLE
		elif (self.slots[0].get_sm() + self.slots[1].get_sm() == self.total_sm) or (self.slots[0].empty_() is False and self.slots[1].empty_() is False):
			return gpu_status_t.FULL
		else:
			return gpu_status_t.ACTIVE

	def get_status_tp(self, tp: float) -> gpu_status_t:
		if self.slots[0].empty_tp(tp) and self.slots[1].empty_tp(tp):
			return gpu_status_t.IDLE
		elif (self.slots[0].get_sm_tp(tp) + self.slots[1].get_sm_tp(tp) == self.total_sm) or (self.slots[0].empty_tp(tp) is False and self.slots[1].empty_tp(tp) is False):
			return gpu_status_t.FULL
		else:
			return gpu_status_t.ACTIVE

	def get_act_sm(self) -> int:
		return self.slots[0].get_sm() + self.slots[1].get_sm()

	def get_rem_sm(self) -> int:
		return self.total_sm - self.get_act_sm()

	def get_act_sm_tp(self, tp: float) -> int:
		self.slots[0].get_sm_tp(tp) + self.slots[1].get_sm_tp(tp)

	def add_sm_into_use(self, _sm: int):
		self.act_sm += _sm

	def rm_sm_from_use(self, _sm: int):
		self.act_sm -= _sm

	def assign_slot(self, _job: Job, sid: int) -> int:
		if sid != -1:
			if self.slots[sid].empty_():
				self.slots[sid].assign_job_slot(_job)
				self.add_sm_into_use(_job.get_sm())
				return sid
			else:
				print("Warning! Assigning an anavailable GPU slot! Automatically select one instead.")
		for i in range(partition_num):
			if self.slots[i].empty_():
				self.slots[i].assign_job_slot(_job)
				self.add_sm_into_use(_job.get_sm())
				return i

	def assign_slot_tp(self, _job: Job, tp: float) -> int:
		for i in range(partition_num):
			if self.slots[i].empty_tp(tp):
				self.slots[i].push_to_waitlist(_job)
				return i

	def reset_slot(self, sid: int):
		self.rm_sm_from_use(self.slots[sid].get_sm())
		self.slots[sid].delete_job_slot()

	def reset_slot_waitlist(self, sid: int): # todo: can it be merged into reset_slot()?
		self.slots[sid].clear_waitlist()
	
	def get_active_slot(self) -> int:
		if self.get_status() == gpu_status_t.ACTIVE:
			if self.slots[0].empty_() is False:
				return 0
			elif self.slots[1].empty_() is False:
				return 1
	
	def get_faster_slot(self) -> int: # only valid when the GPU is full
		sid = -1
		if self.get_status() == gpu_status_t.FULL:
			if self.slots[0].empty_() is False and self.slots[1].empty_() is False:
				if self.slots[0].job.gpu_end_time < self.slots[1].job.gpu_end_time:
					sid = 0
				else:
					sid = 1
			elif self.slots[0].empty_() is False:
				sid = 0
			elif self.slots[1].empty_() is False:
				sid = 1
		return sid

	def calc_power_static(self) -> float:
		return self.power_static
	
	def calc_power_idle(self) -> float:
		if self.get_status() == gpu_status_t.IDLE:
			return 0
		else:
			return (self.max_sm - self.act_sm) * self.power_idle # todo: fix in hw v1 code

	def calc_power_total(self) -> float:
		if self.get_status() == gpu_status_t.IDLE:
			return self.calc_power_static()
		else:
			power = self.calc_power_static() + self.calc_power_idle()
			for i in range(partition_num):
				power += self.slots[i].calc_power_d(self.power_dynamic, self.slots[i].job)
			return power
	
	def get_sm_slot_tp(self, tp: float, sid: int) -> int:
		return self.slots[sid].get_sm_tp(tp)

	def get_act_sm_tp(self, tp: float) -> int:
		ret = 0
		for i in range(partition_num):
			ret += self.slots[i].get_sm_tp(tp)
		return ret


